text/2020-12-16-pd-rbac.md

```diff
@@ -0,0 +1,150 @@
+  - Query users & roles in memory cache.
+  - Load them from KV.
+  - Persist them into KV.
+  - Validate user hash & permission.
+- Add administrative APIs
+  - GET /user
+    - Query user info of current request.
+    - `{"user": "blabla", "roles": ["reader", …]}`
+  - GET /users
+    - List all usernames and their roles.
+  - POST /users
+    - Create a new user.
+    - `{"username": "blabla", "roles": []}`
+  - GET /users/`<user>`
+    - Query user info
+    - `{"username": "blabla", "roles": ["reader", …]}`
+  - POST /users/`<user>`/password
+    - Edit user password.
+    - `{"password": "blabla"}`
+  - POST /users/`<user>`/roles
+    - Set user roles.
+    - `{"roles": ["reader", "writer", …]}`
+  - POST /users/`<user>`/roles/add
+    - Add a role to user.
+    - `{"role": "admin"}`
+  - DELETE /users/`<user>`/roles
+    - Delete a role from user.
+    - `{"role": "admin"}`
+  - GET /roles
+    - List all usernames and their roles.
+    - `[{"name": "reader", "permissions": [{"resource": "Operators", "action": "GET"},…]},…]`
+  - POST /roles
+    - Create a new role.
+    - `{"name": "writer", "permissions": [{"resource": "Operators", "action": "UPDATE"},…]}`
+  - GET /roles/`<role>`
+    - Query role info
+    - `{"name": "reader", "permissions": [{"resource": "Operators", "action": "GET"},…]}`
+  - POST /roles/`<role>`
+    - Set role permissions.
+    - `{"permissions": […]}`
+  - POST /roles/`<role>`/add
+    - Add a permission to role.
+    - `{"permission": {...}}`
+  - DELETE /roles/`<role>`
+    - Delete a permission from role.
+    - `{"permission": {...}}`
+- Add middlewares to HTTP router and gRPC handlers
+  - Rejects any request without a valid pair of username & password.
+  - Maps requested handler to a `Permission`.
+  - Rejects the request if the current user doesn’t have the required permission.
+- Modify pd-ctl, TiKV and TiDB to send requests with user credentials.
+
+### How Clients Interact with PD
+
+- The administrator deploys PD cluster and creates initial users with TiDB/TiKV
+  specific permissions (there are pre-defined roles for TiDB and TiKV).
+- TiDB/TiKV may access PD with these users.
+- pd-ctl and other utils may access PD with the admin user or any existing user.
+  Logging with an unprivileged user may restrict actions these tools may perform.
+
+## Alternatives
+
+### Certificate & Token-based PD Auth
+
+This method is mentioned in [rfcs/46](https://github.com/tikv/rfcs/pull/46).
+This implementation is great generally, but it lacks the ability to have fine-grained
+control on PD APIs.
+
+The idea of authenticating with certificates and its design of tokens is great though.
+We may combine it with the design proposed with this RFC in the future by making
+minor modifications to the administrative APIs and data models, and implementing
+a token issuing and refreshing mechanism.
+
+## Future Additions
+
+- Token-based authentication.
```

Comment:
token-based authorization is considerably more secure than using username/password and IMO should be the only mechanism supported. There are fairly standard approaches which can be used, e.g., OAuth

Reply:
Maybe different users will have different security models? I'm thinking about providing several auth methods, including password-based, token-based, and certificate-based ones.

Reply:
From a security perspective, the more auth methods there are, the less secure your system is. Could we support a single, token-based auth method and provide token generation from username/password in the client?